### PR TITLE
Add properties APIs

### PR DESCRIPTION
**Pull request by danewalton**

### CODE REVIEW

**Comment by danewalton** [comment]
**Note to self make sure to add files to CMake once the json is merged**

**Review by hihigupt — COMMENTED** [review]
is it possible to share the line coverage of these ut test... just to make sure that we are covering > 80% of lines

File: source/azure_iot_hub_client_properties.c [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+ */
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "azure_iot_hub_client.h"
+#include "azure_iot_json_reader.h"
+#include "azure_iot_json_writer.h"
+
+#include "azure_iot_hub_client_properties.h"
```
[the remark this answers is not in the document]
Comment (danewalton) [reply source/azure_iot_hub_client_properties.c:16]: do we want this to be the only header for the implementation and have additional headers be only extra dependencies which the impl might need? (here and elsewhere)

File: source/azure_iot_hub_client_properties.c [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_span xComponentSpan;
+
+    if( ( pxAzureIoTHubClient == NULL ) || ( pxJSONWriter == NULL ) || ( pucComponentName == NULL ) || ( usComponentNameLength == 0 ) )
+    {
+        return eAzureIoTErrorInvalidArgument;
+    }
+
+    xComponentSpan = az_span_create( ( uint8_t * ) pucComponentName, usComponentNameLength );
```
[the remark this answers is not in the document]
Comment (danewalton) [reply source/azure_iot_hub_client_properties.c:34]: I have to check to make sure the pointer isn't `NULL` first. We can either have the split of declaration and assignment or have it all after the precond checks

File: source/azure_iot_hub_client_properties.c [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_iot_hub_client_properties_response_type xCoreResponseType;
+
+    if( ( pxAzureIoTHubClient == NULL ) || ( pxJSONReader == NULL ) ||
+        ( ( xResponseType != eAzureIoTHubPropertiesGetMessage ) && ( xResponseType != eAzureIoTHubPropertiesWriteablePropertyMessage ) ) ||
+        ( pulVersion == NULL ) )
+    {
+        AZLogError( ( "Invalid parameter: xResponseType must be eAzureIoTHubPropertiesGetMessage or eAzureIoTHubPropertiesWriteablePropertyMessage" ) );
```
[the remark this answers is not in the document]
Comment (danewalton) [reply source/azure_iot_hub_client_properties.c:156]: darn yup we should

**Comment by danewalton** [comment]
[Code cov for these stands at 82%](https://github.com/Azure/azure-iot-middleware-freertos/pull/133/checks?check_run_id=2741500904#step:8:141)

File: source/azure_iot_hub_client_properties.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (hihigupt) [reply source/azure_iot_hub_client_properties.c:9]: yup

File: source/azure_iot_hub_client_properties.c [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+                                                                   AzureIoTJSONReader_t * pxJSONReader,
+                                                                   AzureIoTHubMessageType_t xResponseType,
+                                                                   uint32_t * pulVersion )
+{
+    AzureIoTResult_t xResult;
+    az_result xCoreResult;
+    az_iot_hub_client_properties_response_type xCoreResponseType;
+
+    if( ( pxAzureIoTHubClient == NULL ) || ( pxJSONReader == NULL ) ||
+        ( ( xResponseType != eAzureIoTHubPropertiesGetMessage ) && ( xResponseType != eAzureIoTHubPropertiesWriteablePropertyMessage ) ) ||
```
Comment (hihigupt) [inline source/azure_iot_hub_client_properties.c:154]: nit: newline

**Review by hihigupt — APPROVED** [review]
:shipit:

File: source/azure_iot_hub_client_properties.c [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+    {
+        AZLogError( ( "AzureIoTHubClientProperties_GetNextComponentProperty failed: invalid argument" ) );
+        return eAzureIoTErrorInvalidArgument;
+    }
+
+    xComponentSpan = az_span_create( ( uint8_t * ) *ppucComponentName, ( int32_t ) *pulComponentNameLength );
+    xCoreResponseType = xResponseType == eAzureIoTHubPropertiesGetMessage ?
+                        AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_GET : AZ_IOT_HUB_CLIENT_PROPERTIES_RESPONSE_TYPE_DESIRED_PROPERTIES;
+
+    if( az_result_failed(
```
Comment (jspaith) [inline source/azure_iot_hub_client_properties.c:204]: The underlying embedded API is going to return `AZ_ERROR_IOT_END_OF_PROPERTIES` when there's no more properties.  However this is going to end up masking this legit status into an `eAzureIoTErrorFailed`.  I think we need a new status code for `AzureIoTResult_t` and to special case transforming that specific error back to caller.

**Review by jspaith — CHANGES_REQUESTED** [review]

File: source/azure_iot_hub_client_properties.c (hunk 5 above)
Comment (danewalton) [reply source/azure_iot_hub_client_properties.c:204]: Really good catch thank you. Updated

Comment (danewalton) [reply source/azure_iot_hub_client_properties.c:204]: Added to the ut as well to make sure we hit that:
https://github.com/Azure/azure-iot-middleware-freertos/pull/135/commits/0d00447bb19ec12fd1dafa976a81a948c553312b

**Review by jspaith — APPROVED** [review]

**Pull request merged**